README.md
# e3-mrfioc2

## Check the PCI Address of MRF HW
One can find the correct PCI Address by `lspci` command, and e3 also provide a wrapper script for this. 
```
$ bash tools/get_pciaddr.bash 
Usage: get_pciaddr.bash possible_devices [cpcievr220|cpcievr230|cpcievr300|pcieevr300
                                          |mtcaevr300|cpcievg220|cpcievg230|
										  cpcievg300|mtcaevm300] 

```
For example, if one has one MRF PCIE EVR 300DC, one can find 
```sh
$ bash tools/get_pciaddr.bash pcieevr300
1 : Input pcieevr300 ---- PCI ID    03:00.0
```

## Kernel module (mrf.ko) can be installed via DKMS


```sh
$ make init
$ make dkms_add
$ make dkms_build
$ make dkms_install
```

In order to remove them

```sh
$ make dkms_uninstall
$ make dkms_remove
```

## Kernel modules configuration

* Create and load udev rule in /etc/udev/rules.d/99-mrfioc2.rules
* Create and load the autoload configuration in /etc/modules-load.d/mrf.conf
* Remove and load the kernel module with modprobe

```sh
$ make setup
```

In order to clean the configuration,

```sh
$ make setup_clean
```

## Notice
If one has already the running dkms.service in systemd, the next reboot with new kernl image will make the kernel module be ready. However, if one doesn't have one, please run bash dkms/dkms_setup.bash in order to enable dkms.service.

```
$ bash dkms/dkms_setup.bash
$ systemctl status dkms
● dkms.service - Builds and install new kernel modules through DKMS
   Loaded: loaded (/etc/systemd/system/dkms.service; enabled; vendor preset: ena
   Active: active (exited) since Sun 2018-07-29 01:13:59 CEST; 4s ago
     Docs: man:dkms(8)
  Process: 3271 ExecStart=/bin/sh -c dkms autoinstall --verbose --kernelver $(un
 Main PID: 3271 (code=exited, status=0/SUCCESS)


```


CHANGELOG.md
# Changelog


## [Unreleased]

### New Features
- mtca and ts environment added.
- evgr.iocsh added.
- Added evgasynr.iocsh
- Generic evg.iocsh file added.
- 3U chassis added to mtca environment.
- evm.cmd added for the deployment inventory call.


### Bug Fixes
- Fixed D macro


### Refactoring
- evm.iocsh for the ESS naming adjustment.
- Changed SYS, D macros to P, R macros
- No need for macro in embedded EVRs OBJ
- Updated iocsh/evm.iocsh with new macros
- ts.cmd master events moved to "reftabs".



### Documentation Changes
- License definition added.
- Pull Request Template updated to follow the changelog standard.





## Version 7.0.3.1-3.1.2/R-2.2.0-rc8-a1baa66-2002031236 (2020-02-03)









## Version 7.0.3.1-3.1.2/R-2.2.0-rc7-aa11391-1911121618 (2019-11-11)









## Version 7.0.3-3.1.2/R-2.2.0-rc7-2f412fa-1910251151 (2019-10-21)









## Version 7.0.3-3.1.1/R-2.2.0-rc6-c2dc739-1909271722 (2019-09-27)









## Version 2.2.0-rc5 (2019-05-13)









## Version 3.15.5-3.0.2/2.2.0-rc4-01923be-1811101812 (2018-11-09)









## Version 3.15.5-3.0.2/2.2.0-rc2-1aee5fb-1811091351 (2018-11-05)









## Version 3.15.5-3.0.0/2.2.0-rc2-1811020226 (2018-10-06)









## Version R0.2.0.0 (2018-07-31)









## Version R0.1.0.0 (2018-03-21)











tools/get_pciaddr.bash
#!/bin/bash
#
#  Copyright (c) 2019  Jeong Han Lee
#  Copyright (c) 2019  European Spallation Source ERIC
#
#  The program is free software: you can redistribute
#  it and/or modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation, either version 2 of the
#  License, or any newer version.
#
#  This program is distributed in the hope that it will be useful, but WITHOUT
#  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
#  more details.
#
#  You should have received a copy of the GNU General Public License along with
#  this program. If not, see https://www.gnu.org/licenses/gpl-2.0.txt
#
#
#   author  : Jeong Han Lee
#   email   : jeonghan.lee@gmail.com
#   date    : Monday, October 21 13:49:21 CEST 2019
#   version : 0.0.3

EXIST=1
NON_EXIST=0

declare -gr SC_SCRIPT="$(realpath "$0")"
declare -gr SC_TOP="${SC_SCRIPT%/*}"
declare -ga list=()
declare -i i=0; 

function checkIfVar()
{

    local var=$1;shift;
    local result=""
    if [ -z "$var" ]; then
	result=$NON_EXIST
	# doesn't exist
    else
	result=$EXIST
	# exist
    fi
    echo "${result}"
}

## if [[ $(checkIfFile "${release_file}") -eq "$NON_EXIST" ]]; then
#   NON_EXIT
## fi

function checkIfFile
{
    local file=$1; shift;
    local result=""
    if [ ! -e "$file" ]; then
	result=$NON_EXIST
	# doesn't exist
    else
	result=$EXIST
	# exist
    fi
    echo "${result}"
};


CPICEVR220="10dc"
CPCIEVR230="10e6"
CPCIEVR300="152c"
PCIEEVR300="172c"
MTCAEVR300="132c"

CPCIEVG220="20dc"
CPCIEVG230="20e6"
CPCIEVG300="252c"
MTCAEVM300="232c"


# logic test, dummy ids
#CPCIEVR230="2166"
#CPCIEVG230="3b64"

EVR_CPCI_220="cpcievr220"
EVR_CPCI_230="cpcievr230"
EVR_CPCI_300="cpcievr300"
EVR_PCIE_300="pcieevr300"
EVR_MTCA_300="mtcaevr300"
EVG_CPCI_220="cpcievg220"
EVG_CPCI_230="cpcievg230"
EVG_CPCI_300="cpcievg300"
EVM_MTCA_300="mtcaevm300"

options+=${EVR_CPCI_220}
options+="|"
options+=${EVR_CPCI_230}
options+="|"
options+=${EVR_CPCI_300}
options+="|"
options+=${EVR_PCIE_300}
options+="|"
options+=${EVR_MTCA_300}
options+="|"
options+=${EVG_CPCI_220}
options+="|"
options+=${EVG_CPCI_230}
options+="|"
options+=${EVG_CPCI_300}
options+="|"
options+=${EVM_MTCA_300}



case "$1" in

    ${EVR_CPCI_220})
	DEVICEID=${CPICEVR220}
	;;
    ${EVR_CPCI_230})
	DEVICEID=${CPICEVR230}
	;;
    ${EVR_CPCI_300})
	DEVICEID=${CPICEVR300}
	;;
    ${EVR_PCIE_300})
	DEVICEID=${PCIEEVR300}
	;;
    ${EVR_MTCA_300})
	DEVICEID=${MTCAEVR300}
	;;
    ${EVG_CPCI_220})
	DEVICEID=${CPICEVG220}
	;;
    ${EVG_CPCI_230})
	DEVICEID=${CPCIEVG230}
	;;
    ${EVG_CPCI_300})
	DEVICEID=${CPICEVG300}
	;;
    ${EVM_MTCA_300})
	DEVICEID=${MTCAEVM300}
	;;
     *)
	echo "Usage: ${0##*/} possible_devices [${options}] "
	exit 1;
	;;

esac

echo ${DEVICEID}

list="$(lspci -nm |grep ${DEVICEID} | cut -d' ' -f1)"


for rep in ${list[@]}; do
    i+=1;
    printf "%d : Input %10s ---- PCI ID %10s\n" $i $1 $rep
done

exit

